### PR TITLE
[mlir][EmitC] Add `verbatim` op

### PR DESCRIPTION
**Pull request by simon-camp**
The `verbatim` operation produces no results and the value is emitted as is followed by a line break  ('\n' character) during translation.

Note: Use with caution. This operation can have arbitrary effects on the semantics of the emitted code. Use semantically more meaningful operations whenever possible. Additionally this op is *NOT* intended to be used to inject large snippets of code.

This operation can be used in situations where a more suitable operation is not yet implemented in the dialect or where preprocessor directives interfere with the structure of the code.

Co-authored-by: Marius Brehler <marius.brehler@iml.fraunhofer.de>

### CODE REVIEW

File: mlir/test/Dialect/EmitC/invalid_ops.mlir [hunk 1]
```diff
@@ -289,3 +289,11 @@ func.func @test_expression_multiple_results(%arg0: i32) -> i32 {
...
   return %r : i32
 }
+
+// -----
+
+func.func @verbatim_trailing_semicolon() {
+    // expected-error @+1 {{'emitc.verbatim' op ';' not allowed as the last character, use the `trailing_semicolon` attribute instead}}
+    emitc.verbatim "typedef int32_t i32;"
+    return
+}
```
Comment (marbre) [inline mlir/test/Dialect/EmitC/invalid_ops.mlir:299]: Missing newline at EOF.

File: mlir/test/Dialect/EmitC/ops.mlir [hunk 2]
```diff
@@ -166,3 +166,14 @@ func.func @test_for_not_index_induction(%arg0 : i16, %arg1 : i16, %arg2 : i16) {
...
+emitc.verbatim "#ifdef __cplusplus"
+emitc.verbatim "extern \"C\" {"
+emitc.verbatim "#endif  // __cplusplus"
+
+emitc.verbatim "#ifdef __cplusplus"
+emitc.verbatim "}  // extern \"C\""
+emitc.verbatim "#endif  // __cplusplus"
+
+emitc.verbatim "typedef int32_t i32" {trailing_semicolon = unit}
+emitc.verbatim "typedef float f32" trailing_semicolon
```
Comment (marbre) [inline mlir/test/Dialect/EmitC/ops.mlir:179]: Do you want to add one of those or even both to the example in the description?

**Review by marbre — CHANGES_REQUESTED** [review]
Thanks @simon-camp :) As we've already discussed this offline for quite a while, I don't have much to say. From a first quick review, there are only minor things to change.

**Comment by AlexDenisov** [comment]
This is very useful!

Out of curiosity, what makes the `trailing_semicolon` necessary? Isn't it easier to include `;` into the string itself, and the overcome the validation by adding a space at the end of the string?

**Comment by jpienaar** [comment]
> This is very useful!
> 
> Out of curiosity, what makes the `trailing_semicolon` necessary? Isn't it easier to include `;` into the string itself, and the overcome the validation by adding a space at the end of the string?

+1, could the attribute be avoided?

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td [hunk 3]
```diff
@@ -544,6 +544,43 @@ def EmitC_VariableOp : EmitC_Op<"variable", []> {
   let hasVerifier = 1;
 }
 
+def EmitC_VerbatimOp : EmitC_Op<"verbatim"> {
+  let summary = "Verbatim operation";
+  let description = [{
+    The `verbatim` operation produces no results and the value is emitted as is
+    followed by a '\n' character during translation.
```
Comment (marbre) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:551]:
```suggestion
   followed by a line break  '\n' character during translation.
```

**Comment by simon-camp** [comment]
> This is very useful!
> 
> Out of curiosity, what makes the `trailing_semicolon` necessary? Isn't it easier to include `;` into the string itself, and the overcome the validation by adding a space at the end of the string?

Good suggestion, thank you.

I'm not sure I follow the validation part. The semicolon is now part of the attribute, so there is nothing to verify. Nonetheless this made me think of the implicit `\n` generated during emission, so this is now mentioned in the op description.

**Review by marbre — APPROVED** [review]

**Pull request merged**